File: prob1.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes


class Solution {
    public int partition(int nums[],int si,int ei){
        int pivot=nums[ei];
        int i=si-1;
        for(int j=si;j<ei;j++){
            if(nums[j]<=pivot){
                i++;
                int temp=nums[i];
                nums[i]=nums[j];
                nums[j]=temp;
            }
        }
        i++;
        int temp=nums[ei];
        nums[ei]=nums[i];
        nums[i]=temp;
        return i;
    }
    public int quickSelect(int nums[],int si,int ei,int k){
        int pi=partition(nums,si,ei);
        if(k<pi){
            return quickSelect(nums,si,pi-1,k);
        } else if(k>pi){
            return quickSelect(nums,pi+1,ei,k);
        } else {
            return nums[pi];
        }
    }
    public int findKthLargest(int[] nums, int k) {
        if(nums.length==1){
            return nums[0];
        }
        k=nums.length-k;
        return quickSelect(nums,0,nums.length-1,k);
    }
}

File: prob1.py
# Time Complexity: O(nlogk) using min heap
# Space Complexity: O(n) for heap as we insert all elements into the heap
# Did this run successfully on Leetcode: Yes
# Any problem while coding this question: Faced issues with max heap implementation
# Approach: using min heap, remove from min heap till count > k, 
# at that point, the next element popped will be the kth largest
# for max  heap, remove from max heap, till count > len(nums) - k, 
# at that point, the next element popped will be the kth largest


# using min heap
import heapq

class Solution:

    def findKthLargest(self, nums: List[int], k: int) -> int:
        
        heapq.heapify(nums)
        count = len(nums)
        while count > k:
            heapq.heappop(nums)
            count -= 1
        return heapq.heappop(nums)

# using max heap
# import heapq
# class Solution:
#     def findKthLargest(self, nums: List[int], k: int) -> int:
        
#         for num in nums:
#             num *= -1
            
#         li = []
#         for num in nums:
#             heapq.heappush(li, num)
        
#         for i in range(len(nums) - k):
#             heapq.heappop(li)
        
#         return heapq.heappop(li)